Tdd tut #155730453

MidWeekUser opened the pull request:
## Description
This PR is for learning how to use tests when creating APIs.
## Tasks Completed:
- Created folder structure.
- Created test files in 'tests' folder.
- Ran failing tests to verify of the run.
- Created simple API apps in 'api' folder.
- Ran tests to verify the code.
- Debugged errors to run successful tests.
## Manual Testing (using Bash)
- Download all files to a directory
- Install requirements in 'requirements_flusk_restfulpus.txt' file
- Navigate to root folder
- Use pytest to run the test files
## Pivotal Tracker
Click [here](https://www.pivotaltracker.com/story/show/155730453) to see the story relevant to this PR.

MidWeekUser committed:
[Chore #155730453] Adding new files for tests
- Created new user code file in api folder
- Created new usertests file in tests folder

MidWeekUser committed:
[Chore #155730453] Last commit didn't contain usertests file as indicated. This corrects that

georgreen commented on app/api/users.py:
```diff
@@ -0,0 +1,15 @@
+
+app = Flask(__name__)
+api = Api(app)
+# @api.route('/hello')
```
Remove comments from your work before you push.

MidWeekUser replied:
Will do...

MidWeekUser committed:
[Chore #155730453] Final Test Files For Tutorial
- Create proper API structure
- Create user management api in api folder
- Add model to models file in app folder
- Create test for user management in tests folder
[Finishes #155730453]

georgreen commented on app/api/user_management.py:
```diff
@@ -0,0 +1,56 @@
+    
+    def post(self):
+        
+        user_data = request.json
```
You are not validating the data received from the client. Please try to incorporate this.

georgreen commented on app/api/user_management.py:
```diff
@@ -0,0 +1,56 @@
+        email = user_data['email']
+        firstname = user_data['firstname']
+        secondname = user_data['secondname']
+        username = user_data['username']
```
This code is unsafe and prone to bugs, accessing dictionary using the square bracket can throw exceptions when the key accessed is missing. Correct this by checking the key exists in the dictionary.

georgreen commented on app/api/user_management.py:
```diff
@@ -0,0 +1,56 @@
+        
+        return status
+    
+    def get(self):
```
What is a get Verb for in a register resource?

georgreen commented on app/api/user_management.py:
```diff
@@ -0,0 +1,56 @@
+    
+
+# class Login(Resource):
+#     from app.models import UserManagement
```
You should remove unused code from your work before you push.

georgreen commented on app/tests/test_user_management.py:
```diff
@@ -0,0 +1,25 @@
+class user_scenarios(unittest.TestCase):
+    def setUp(self):
+        self.postman = app.test_client()
+        self.header = {'Content-type': 'application/json'}
```
This is incorrect, setting the content-type should not be done in the header.

georgreen commented on run.py:
```diff
@@ -1,6 +1,8 @@
-from app import app
+from app import app_factory
+
+app = app_factory('DevelopmentConfig')
```
You should not hard code environments, this should be picked from the shell environment. This makes your app more flexible.

georgreen reviewed (commented):
You should not push your virtual env to git, please rectify this.

MidWeekUser committed:
[Chore #155730453] Final changes
- Correct unused code issue
- Remove virtual environment from repo
- Validating data recieved
[Finishes #155730453]

MidWeekUser committed:
[Chore #155730453] Adding missong files to api
[Finishes #155730453]

MidWeekUser committed:
[Chore #155730453] Removing unnecessary files
[Finishes #155730453]

MidWeekUser committed:
git commit -m"[Chore #155730453] Modifes user management
[Finishes #155730453]
